src/include/miopen/graphapi/graphapi_operation.hpp

```diff
@@ -0,0 +1,131 @@
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#pragma once
+
+#include <miopen/errors.hpp>
+#include <miopen/graphapi/graphapi_tensor.hpp>
+
+#include <algorithm>
+#include <tuple>
+#include <vector>
+
+namespace miopen {
+
+namespace graphapi {
+
+class Operation
+{
+public:
+    enum class ArgumentId
+    {
+        X,
+        Y,
+        W
+    };
+
+    Operation& clearInputsOutputs()
+    {
+        mInputs.clear();
+        mOutputs.clear();
+        return *this;
+    }
+    Operation& setInput(ArgumentId argumentId, Tensor* pTensor)
+    {
+        auto existing =
+            std::find_if(mInputs.begin(), mInputs.end(), [argumentId](const auto& item) {
+                return std::get<0>(item) == argumentId;
+            });
+        if(existing == mInputs.end())
+        {
+            mInputs.emplace_back(argumentId, pTensor);
+        }
+        else
+        {
+            std::get<1>(*existing) = pTensor;
+        }
+        return *this;
+    }
+    Operation& setOutput(ArgumentId argumentId, Tensor* pTensor)
+    {
+        auto existing =
+            std::find_if(mOutputs.begin(), mOutputs.end(), [argumentId](const auto& item) {
+                return std::get<0>(item) == argumentId;
+            });
+        if(existing == mOutputs.end())
+        {
+            mOutputs.emplace_back(argumentId, pTensor);
+        }
+        else
+        {
+            std::get<1>(*existing) = pTensor;
+        }
+        return *this;
+    }
+    Tensor* getInput(ArgumentId argumentId) const
+    {
+        auto searched =
+            std::find_if(mInputs.begin(), mInputs.end(), [argumentId](const auto& item) {
+                return std::get<0>(item) == argumentId;
+            });
+        if(searched == mInputs.end())
+        {
+            MIOPEN_THROW(miopenStatusInternalError);
+        }
+        return std::get<1>(*searched);
+    }
+    Tensor* getOutput(ArgumentId argumentId) const
+    {
+        auto searched =
+            std::find_if(mOutputs.begin(), mOutputs.end(), [argumentId](const auto& item) {
+                return std::get<0>(item) == argumentId;
+            });
+        if(searched == mOutputs.end())
+        {
+            MIOPEN_THROW(miopenStatusInternalError);
+        }
+        return std::get<1>(*searched);
+    }
+    bool hasInput(Tensor* pTensor)
+    {
+        return std::find_if(mInputs.begin(), mInputs.end(), [pTensor](const auto& item) {
+                   return std::get<1>(item) == pTensor;
+               }) != mInputs.end();
+    }
+    bool hasOutput(Tensor* pTensor)
+    {
+        return std::find_if(mOutputs.begin(), mOutputs.end(), [pTensor](const auto& item) {
+                   return std::get<1>(item) == pTensor;
+               }) != mOutputs.end();
+    }
+
+private:
+    std::vector<std::tuple<ArgumentId, Tensor*>> mInputs;
+    std::vector<std::tuple<ArgumentId, Tensor*>> mOutputs;
```

Comment:
prefer `std::pair` to `std::tuple` and you should still be able to use `std::get<>`. `pair` has a simple layout while `tuple` doesn't.

Reply:
agree, `pair` will do